gen3-client/g3cmd/download-multiple.go

```diff
@@ -17,7 +19,110 @@ import (
 	"github.com/uc-cdis/gen3-client/gen3-client/jwt"
 )
 
-func batchDownload(numParallel int, reqs []*grab.Request) {
+func lastString(ss []string) string {
+	return ss[len(ss)-1]
+}
+
+func processOriginalFilename(downloadPath string, actualFilename string) string {
+	_, err := os.Stat(downloadPath + actualFilename)
+	if os.IsNotExist(err) {
+		return actualFilename
+	}
+	extension := filepath.Ext(actualFilename)
+	filename := strings.TrimSuffix(actualFilename, extension)
+	counter := 2
+	for {
+		newFilename := filename + "_" + strconv.Itoa(counter) + extension
+		_, err := os.Stat(downloadPath + newFilename)
+		if os.IsNotExist(err) {
+			return newFilename
+		}
+		counter++
+	}
+}
+
+func processS3URLForFilename(presignedURL string, guid string, downloadPath string, filenameFormat string, overwrite bool, renamedFiles *[]RenamedFileInfo) string {
+	if filenameFormat != "guid" {
+		urlWithFilename := strings.Split(presignedURL, "?")[0]
+		actualFilename := lastString(strings.Split(urlWithFilename, guid+"/"))
+		if actualFilename != "" {
+			if filenameFormat == "original" {
+				if !overwrite {
+					newFilename := processOriginalFilename(downloadPath, actualFilename)
+					if actualFilename != newFilename {
+						*renamedFiles = append(*renamedFiles, RenamedFileInfo{GUID: guid, OldFilename: actualFilename, NewFilename: newFilename})
+					}
+					return newFilename
+				}
+				return actualFilename
+			} else if filenameFormat == "combined" {
+				return guid + "_" + actualFilename
+			}
+		}
+	}
+	return guid
+}
+
+func validateFilenameFormat(downloadPath string, filenameFormat string, overwrite bool) {
+	if filenameFormat != "original" && filenameFormat != "guid" && filenameFormat != "combined" {
+		log.Fatalln("Invalid option found! Option \"filename-format\" can either be \"original\", \"guid\" or \"combined\" only")
+	}
+	if filenameFormat == "guid" || filenameFormat == "combined" {
+		fmt.Printf("WARNING: in \"guid\" or \"combined\" mode, duplicated files under \"%s\" will be overwritten!\n", downloadPath)
+		if !commonUtils.AskForConfirmation("Proceed?") {
+			log.Println("Aborted by user")
+			os.Exit(0)
+		}
+	} else if overwrite {
+		fmt.Printf("WARNING: flag \"overwrite\" was set to true in \"original\" mode, duplicated files under \"%s\" will be overwritten!\n", downloadPath)
```

Comment:
you could also print a warning for `overwrite = false` about the counter thing
idk if it's necessary to ask for confirmation in that case

Reply:
added in https://github.com/uc-cdis/cdis-data-client/pull/56/commits/8f46c72e9bb0eba742d8281ca969cded23f6f33a